platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/server/DaemonServices.java

```diff
@@ -136,17 +137,32 @@ protected GarbageCollectorMonitoringStrategy createGarbageCollectorMonitoringStr
         return GarbageCollectorMonitoringStrategy.determineGcStrategy();
     }
 
-    protected ImmutableList<DaemonCommandAction> createDaemonCommandActions(DaemonContext daemonContext, ProcessEnvironment processEnvironment, DaemonHealthStats healthStats, DaemonHealthCheck healthCheck, BuildExecuter buildActionExecuter, DaemonRunningStats runningStats) {
+    protected CleanUpVirtualFileSystemAfterBuild createCleanUpVirtualFileSystemAfterBuild(ExecutorFactory executorFactory, GradleUserHomeScopeServiceRegistry userHomeServiceRegistry) {
+        return new CleanUpVirtualFileSystemAfterBuild(executorFactory, userHomeServiceRegistry);
+    }
+
+    protected ImmutableList<DaemonCommandAction> createDaemonCommandActions(
+        BuildExecuter buildActionExecuter,
+        DaemonContext daemonContext,
+        DaemonHealthCheck healthCheck,
+        DaemonHealthStats healthStats,
+        DaemonRunningStats runningStats,
+        ProcessEnvironment processEnvironment,
+
+        CleanUpVirtualFileSystemAfterBuild cleanUpVirtualFileSystemAfterBuild
+    ) {
         File daemonLog = getDaemonLogFile();
         DaemonDiagnostics daemonDiagnostics = new DaemonDiagnostics(daemonLog, daemonContext.getPid());
+        GradleUserHomeScopeServiceRegistry gradleUserHomeScopeServiceRegistry = get(GradleUserHomeScopeServiceRegistry.class);
         return ImmutableList.of(
             new HandleStop(get(ListenerManager.class)),
-            new HandleInvalidateVirtualFileSystem(get(GradleUserHomeScopeServiceRegistry.class)),
+            new HandleInvalidateVirtualFileSystem(gradleUserHomeScopeServiceRegistry),
             new HandleCancel(),
+            new EstablishBuildEnvironment(processEnvironment),
+            cleanUpVirtualFileSystemAfterBuild,
             new HandleReportStatus(),
             new ReturnResult(),
             new StartBuildOrRespondWithBusy(daemonDiagnostics), // from this point down, the daemon is 'busy'
-            new EstablishBuildEnvironment(processEnvironment),
             new LogToClient(loggingManager, daemonDiagnostics), // from this point down, logging is sent back to the client
             new LogAndCheckHealth(healthStats, healthCheck, runningStats),
```

Comment:
❓ Why did you need to change the order here? What is the effect of changing the order?

Reply:
I needed this to get to the system property specifying the number of allowed watched hierarchies.

This should not be needed anymore, since I split up the cleanup between "before build finished" cleanup (all the existing cleanup, minus symlink cleanup) and "after build finished" cleanup (direct + indirect symlink cleanup).

Reverted.